workflow/util/util.go

```diff
@@ -880,151 +858,437 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
-						// the modified supplied parameter values.
-						if node.Type != wfv1.NodeTypeSuspend {
-							deletedNodes[node.ID] = true
-							deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, node, deletedPods, podsToDelete)
-							log.Debugf("Deleted pod node: %s", node.Name)
-						}
+	failed := make(map[string]bool)
+	for nodeID, node := range wf.Status.Nodes {
+		if node.Phase.FailedOrError() {
+			failed[nodeID] = true
+		}
+	}
 
-						descendantNodeIDs := getDescendantNodeIDs(wf, node)
-						for _, descendantNodeID := range descendantNodeIDs {
-							deletedNodes[descendantNodeID] = true
-							descendantNode, err := wf.Status.Nodes.Get(descendantNodeID)
-							if err != nil {
-								log.Warnf("Was unable to obtain node for %s due to %s", descendantNodeID, err)
-								return nil, nil, fmt.Errorf("Was unable to obtain node for %s due to %s", descendantNodeID, err)
-							}
-							if descendantNode.Type == wfv1.NodeTypePod {
-								newWF, resetParentGroupNodes = resetConnectedParentGroupNodes(wf, newWF, node, resetParentGroupNodes)
-								deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, *descendantNode, deletedPods, podsToDelete)
-								log.Debugf("Deleted pod node %s since it belongs to node %s", descendantNode.Name, node.Name)
-							}
-						}
-					} else {
-						log.Debugf("Reset non-pod/suspend/skipped node %s", node.Name)
-						newNode := node.DeepCopy()
-						newWF.Status.Nodes.Set(newNode.ID, resetNode(*newNode))
-					}
-				}
-			} else {
-				if !containsNode(resetParentGroupNodes, node.ID) {
-					log.Debugf("Node %s remains as is", node.Name)
-					newWF.Status.Nodes.Set(node.ID, node)
-				}
-			}
-		case wfv1.NodeError, wfv1.NodeFailed, wfv1.NodeOmitted:
-			if isGroupNode(node) {
-				newNode := node.DeepCopy()
-				newWF.Status.Nodes.Set(newNode.ID, resetNode(*newNode))
-				log.Debugf("Reset %s node %s since it's a group node", node.Name, string(node.Phase))
-				continue
-			} else {
-				if node.Type != wfv1.NodeTypeRetry && isDescendantNodeSucceeded(wf, node, nodeIDsToReset) {
-					log.Debugf("Node %s remains as is since it has succeed child nodes.", node.Name)
-					newWF.Status.Nodes.Set(node.ID, node)
-					continue
-				}
-				log.Debugf("Deleted %s node %s since it's not a group node", node.Name, string(node.Phase))
-				deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, node, deletedPods, podsToDelete)
-				log.Debugf("Deleted pod node: %s", node.Name)
-				deletedNodes[node.ID] = true
-			}
-			// do not add this status to the node. pretend as if this node never existed.
-		default:
-			// Do not allow retry of workflows with pods in Running/Pending phase
-			return nil, nil, errors.InternalErrorf("Workflow cannot be retried with node %s in %s phase", node.Name, node.Phase)
+	deleteNodes, err := getNodeIDsToResetNoChildren(restartSuccessful, nodeFieldSelector, wf.Status.Nodes)
```

Comment:
Would I have to set the --restart-successful in order to identify the nodes that need to be deleted, even if I only wish to retry a specific failed node?

Reply:
No this wouldn't be the case